internal/test/wsu/wsu_test.go

```diff
@@ -84,37 +86,51 @@ func TestWSU(t *testing.T) {
 	require.NotEmptyf(t, playbookPath, "WSU_PATH environment variable not set")
 	require.NotEmptyf(t, clusterAddress, "CLUSTER_ADDR environment variable not set")
 
-	// In order to run the ansible playbook we create an inventory file:
-	// https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html
-	hostFilePath, err := createHostFile(framework.Credentials.GetIPAddress(), framework.Credentials.GetPassword())
-	require.NoErrorf(t, err, "Could not write to host file: %s", err)
-	cmd := exec.Command("ansible-playbook", "-vvv", "-i", hostFilePath, playbookPath)
-	out, err := cmd.CombinedOutput()
-	require.NoError(t, err, "WSU playbook returned error: %s, with output: %s", err, string(out))
-
-	// Ansible will copy files to a temporary directory with a path such as:
-	// C:\\Users\\Administrator\\AppData\\Local\\Temp\\ansible.z5wa1pc5.vhn\\
-	initialSplit := strings.Split(string(out), "C:\\\\Users\\\\Administrator\\\\AppData\\\\Local\\\\Temp\\\\ansible.")
-	require.True(t, len(initialSplit) > 1, "Could not find Windows temp dir: %s", out)
-	ansibleTempDir = "C:\\Users\\Administrator\\AppData\\Local\\Temp\\ansible." + strings.Split(initialSplit[1], "\"")[0]
-
-	t.Run("Files copied to Windows node", testFilesCopied)
-	t.Run("Pending CSRs were approved", testNoPendingCSRs)
-	t.Run("Node is in ready state", testNodeReady)
-	// test if the Windows node has proper worker label.
-	t.Run("Check if worker label has been applied to the Windows node", testWorkerLabelsArePresent)
-	t.Run("Network annotations were applied to node", testHybridOverlayAnnotations)
-	t.Run("HNS Networks were created", testHNSNetworksCreated)
-	t.Run("Check cni config generated on the Windows host", testCNIConfig)
-	t.Run("East-west networking", testEastWestNetworking)
+	for _, vm := range framework.WinVMs {
+
+		// In order to run the ansible playbook we create an inventory file:
+		// https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html
+		hostFilePath, err := createHostFile(vm.Credentials.GetIPAddress(),
+			vm.Credentials.GetPassword())
+		require.NoErrorf(t, err, "Could not write to host file: %s", err)
+		cmd := exec.Command("ansible-playbook", "-vvv", "-i", hostFilePath, playbookPath)
+		out, err := cmd.CombinedOutput()
+		require.NoError(t, err, "WSU playbook returned error: %s, with output: %s", err, string(out))
```

Comment:
`createHostFile()` can be modified to take the slice of WinVMs, creating a single host file for all VMs. You can then move it, and else in this section of of the for loops, so you only need to run the ansible-playbook once.  

This will save a bunch of time, and does not require goroutines/channels like the rest of the parallelization.

You'll also need to figure out a different solution for gathering the temp directory.

Reply:
This will be handled in a future ticket